boa/vm/py_evm.py

```diff
@@ -0,0 +1,579 @@
+
+        self.vm.patch.timestamp = int(block_info["timestamp"], 16)
+        self.vm.patch.block_number = int(block_info["number"], 16)
+        # TODO patch the other stuff
+
+        self.vm.state._account_db._rpc._init_mem_db()
+
+    @property
+    def is_forked(self):
+        return self.vm.__class__._state_class.account_db_class == AccountDBFork
+
+    def _set_account_db_class(self, account_db_class: type):
+        self.vm.__class__._state_class.account_db_class = account_db_class
+
+    def get_gas_meter_class(self):
+        return self.vm.state.computation_class._gas_meter_class
+
+    def set_gas_meter_class(self, cls: type):
+        self.vm.state.computation_class._gas_meter_class = cls
+
+    def get_balance(self, address: Address):
+        return self.vm.state.get_balance(address.canonical_address)
+
+    def set_balance(self, address: Address, value):
+        self.vm.state.set_balance(address.canonical_address, value)
+
+    def get_code(self, address: Address) -> bytes:
+        return self.vm.state.get_code(address.canonical_address)
+
+    def get_gas_limit(self):
+        return self.vm.state.gas_limit
+
+    # advanced: reset warm/cold counters for addresses and storage
+    def reset_access_counters(self):
+        self.vm.state._account_db._reset_access_counters()
+
+    def snapshot(self) -> Tuple[Hash32, JournalDBCheckpoint]:
+        return self.vm.state.snapshot()
+
+    def anchor(self):
+        return self.vm.patch.anchor()
+
+    def revert(self, snapshot_id: Tuple[Hash32, JournalDBCheckpoint]) -> None:
+        self.vm.state.revert(snapshot_id)
+
+    def generate_contract_address(self, sender: Address):
+        nonce = self.vm.state.get_nonce(sender.canonical_address)
+        self.vm.state.increment_nonce(sender.canonical_address)
+        return Address(generate_contract_address(sender.canonical_address, nonce))
+
+    def deploy_code(
+        self,
+        sender: Address,
+        origin: Address,
+        target_address: Address,
+        prefetch_state: bool,
+        gas: Optional[int],
+        gas_price: int,
+        value: int,
+        bytecode: bytes,
+    ):
+        if gas is None:
+            gas = self.vm.state.gas_limit
+
+        msg = Message(
+            to=constants.CREATE_CONTRACT_ADDRESS,  # i.e., b""
+            sender=sender.canonical_address,
+            gas=gas,
+            value=value,
+            code=bytecode,
+            create_address=target_address.canonical_address,
+            data=b"",
+        )
+
+        if prefetch_state:
+            self.vm.state._account_db.try_prefetch_state(msg)
+
+        tx_ctx = BaseTransactionContext(
+            origin=origin.canonical_address, gas_price=gas_price
+        )
+        return self.vm.state.computation_class.apply_create_message(
+            self.vm.state, msg, tx_ctx
+        )
+
+    def execute_code(
+        self,
+        sender: Address,
+        to: Address,
+        prefetch_state: bool,
```

Comment:
i think prefetch_state should be part of the PyEVM state, it is an implementation detail for fork mode